src/main/java/org/elasticsearch/plugins/repository/DefaultPluginDownloader.java

```diff
@@ -0,0 +1,381 @@
+                        java.security.cert.X509Certificate[] certs, String authType) {
+                }
+
+                public void checkServerTrusted(
+                        java.security.cert.X509Certificate[] certs, String authType) {
+                }
+            }
+    };
+
+    private static final HostnameVerifier TRUST_ALL_HOSTS = new HostnameVerifier() {
+        public boolean verify(String hostname, SSLSession session) {
+            return true;
+        }
+    };
+
+    public DefaultPluginDownloader(Path root, TimeValue timeout, String userAgent, String elasticsearchVersion) {
+        this.root = root;
+        this.userAgent = userAgent;
+        this.elasticsearchVersion = elasticsearchVersion;
+        if (timeout != null) {
+            this.timeout = timeout;
+        }
+    }
+
+    @Override
+    public Path download(String source) throws IOException {
+        Files.createDirectories(root);
+
+        // Install the all-trusting trust manager
+        try {
+            SSLContext sc = SSLContext.getInstance("SSL");
+            sc.init(null, TRUST_ALL_CERTS, new java.security.SecureRandom());
+            HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
+            HttpsURLConnection.setDefaultHostnameVerifier(TRUST_ALL_HOSTS);
+        } catch (GeneralSecurityException e) {
+            throw new ElasticsearchException("Exception when installing the all-trusting trust manager", e);
+        }
+
+        // Destination for the next download
+        Path destination = root.resolve(String.valueOf(downloadNumber.get()));
+        while (Files.exists(destination)) {
+            destination = root.resolve(String.valueOf(downloadNumber.getAndIncrement()));
+        }
+
+        DownloadThread dlThread = new DownloadThread(source, destination, listeners, userAgent, elasticsearchVersion);
+        try {
+            dlThread.setDaemon(true);
+            dlThread.start();
+            dlThread.join(timeout.millis());
+
+            if (dlThread.isAlive()) {
+                throw new ElasticsearchTimeoutException("The download operation took longer than " + timeout + ", stopping it.");
+            }
+        } catch (InterruptedException e) {
+            return null;
+        } finally {
+            dlThread.closeStreams();
+        }
+
+        if (dlThread.wasSuccessful()) {
+            return destination;
+        }
+        return null;
+    }
+
+    @Override
+    public void addListener(Listener listener) {
+        listeners.add(listener);
+    }
+
+
+    private class DownloadThread extends Thread {
+
+        private final String source;
+        private final Path destination;
+        private Collection<Listener> listeners;
+
+        private final String userAgent;
+        private final String elasticsearchVersion;
+
+        private boolean success = false;
+        private Exception exception = null;
+
+        private InputStream input = null;
```

Comment:
You can move `input`, `output` and `connection` in download method. There is no need to have them as fields.